[Driver][SYCL] Add initial SYCL offload compilation support

mdtoguchi opened the pull request:
Introduces the SYCL based toolchain and initial toolchain construction when using the '-fsycl' option.  This option will enable SYCL based offloading, creating a SPIR-V based IR file packaged into the compiled host object.

This includes early support for creating the host/device object using the new offloading model.  The device object is created using the spir64-unknown-unknown target triple.

New/Updated Options:
 -fsycl  Enables SYCL offloading for host and device
 -fsycl-device-only
         Enables device only compilation for SYCL
 -fsycl-host-only
         Enables host only compilation for SYCL

RFC Reference: https://discourse.llvm.org/t/rfc-sycl-driver-enhancements/74092

mdtoguchi committed:
[Driver][SYCL] Add initial SYCL offload compilation support

Introduces the SYCL based toolchain and initial toolchain construction
when using the '-fsycl' option.  This option will enable SYCL based
offloading, creating a SPIR-V based IR file packaged into the compiled
host object.

This includes early support for creating the host/device object using
the new offloading model.  The device object is created using the
spir64-unknown-unknown target triple.

New/Updated Options:
 -fsycl  Enables SYCL offloading for host and device
 -fsycl-device-only
         Enables device only compilation for SYCL
 -fsycl-host-only
         Enables host only compilation for SYCL

RFC Reference: https://discourse.llvm.org/t/rfc-sycl-driver-enhancements/74092

AaronBallman commented on clang/lib/Driver/CMakeLists.txt:
```diff
@@ -85,6 +85,7 @@ add_clang_library(clangDriver
   ToolChains/PPCFreeBSD.cpp
   ToolChains/InterfaceStubs.cpp
   ToolChains/ZOS.cpp
+  ToolChains/SYCL.cpp
```
You should keep this alphabetized

AaronBallman commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,202 @@
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include <algorithm>
+#include <sstream>
```
Why is this include needed?

AaronBallman commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,202 @@
+}
+
+// Unsupported options for SYCL device compilation
+//  -fcf-protection, -fsanitize, -fprofile-generate, -fprofile-instr-generate
+//  -ftest-coverage, -fcoverage-mapping, -fcreate-profile, -fprofile-arcs
+//  -fcs-profile-generate -forder-file-instrumentation, --coverage
```
I expect this comment to get out of date rather quickly; I'd probably drop the list and let users read the function body instead.

tahonermann replied:
Another option would be to add each of the option names in a comment on the same line as the `OPT_` prefixed name below.

AaronBallman commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,202 @@
+//  -fcf-protection, -fsanitize, -fprofile-generate, -fprofile-instr-generate
+//  -ftest-coverage, -fcoverage-mapping, -fcreate-profile, -fprofile-arcs
+//  -fcs-profile-generate -forder-file-instrumentation, --coverage
+static std::vector<OptSpecifier> getUnsupportedOpts(void) {
```
```suggestion
static std::vector<OptSpecifier> getUnsupportedOpts() {
```

tahonermann commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -979,6 +1000,44 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
+
+  llvm::SmallVector<llvm::Triple, 4> UniqueSYCLTriplesVec;
+
+  // If -fsycl is supplied we will assume SPIR-V.
```
Actually, this comment doesn't seem all that helpful here either. Perhaps it would be better colocated with `addSYCLDefaultTriple()`. Hmm, no, perhaps `getDefaultSYCLArch()` then?

tahonermann commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -6553,6 +6634,20 @@ const ToolChain &Driver::getOffloadingDeviceToolChain(
       break;
     }
+    case Action::OFK_SYCL:
+      // TODO: Add additional Arch values for Ahead of Time support for SYCL.
```
Since AoT support doesn't exist in Clang yet, this TODO comment seems premature.

tahonermann commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -5116,15 +5123,39 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
 
-    if (Arg *A = Args.getLastArg(options::OPT_sycl_std_EQ)) {
-      A->render(Args, CmdArgs);
+    // Set O2 optimization level by default
+    if (!Args.getLastArg(options::OPT_O_Group))
+      CmdArgs.push_back("-O2");
```
What is the motivation for defaulting the optimization level to something different than the host? I find this surprising.

mdtoguchi replied:
The original implementation was based off of defaults that were applied for OpenCL:  https://github.com/llvm/llvm-project/blob/main/clang/lib/Frontend/CompilerInvocation.cpp#L697-L700

Had a short interaction with @bader and he stated optimizations by default are critical for SYCL, as it adds a bit of C++ code like template instantiations to user's code.

tahonermann commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -5116,15 +5123,39 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+
+    // Add any options that are needed specific to SYCL offload while
+    // performing the host side compilation.
+    if (!IsSYCLDevice) {
+      // Let the front-end host compilation flow know about SYCL offload
+      // compilation.
+      CmdArgs.push_back("-fsycl-is-host");
+    }
```
This code seems like it should be on the `else` branch of the `if (IsSYCLDevice)` statement at line 5128.

mdtoguchi replied:
Being in an `else` for `IsSYCLDevice` will pass `-fsycl-is-host` to all non-SYCL device compilations, even when offloading isn't enabled.  Maybe this can be structured to all be in an `IsSYCL` condition.  I will look into that.

tahonermann commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,202 @@
+  // Add the SYCL header search locations in the specified order.
+  //   ../include/sycl
+  //   ../include/sycl/stl_wrappers
+  //   ../include
```
Why add `../include`? That doesn't seem SYCL specific.

mdtoguchi replied:
The current location used for the SYCL based headers is not a standard location.  Scanning through some existing code, simple inclusion of the headers is something to the effect of `#include <sycl/sycl.hpp>`.  So, this allows for capturing of the headers as needed.

tahonermann commented on clang/lib/Driver/ToolChains/SYCL.h:
```diff
@@ -0,0 +1,82 @@
+  bool useIntegratedAs() const override { return true; }
+  bool isPICDefault() const override { return false; }
+  llvm::codegenoptions::DebugInfoFormat getDefaultDebugFormat() const override {
+    if (this->HostTC.getTriple().isWindowsMSVCEnvironment())
+      return this->HostTC.getDefaultDebugFormat();
+    return ToolChain::getDefaultDebugFormat();
```
When/why would the host debug format not be what is desired?

mdtoguchi replied:
Good point, this can probably be cleaned up.

tahonermann commented on clang/lib/Driver/ToolChains/SYCL.h:
```diff
@@ -0,0 +1,82 @@
+
+  SanitizerMask getSupportedSanitizers() const override;
+
+  const ToolChain &HostTC;
+
+  SYCLInstallationDetector SYCLInstallation;
```
```suggestion
private:
  const ToolChain &HostTC;
  SYCLInstallationDetector SYCLInstallation;
```

tahonermann commented on llvm/include/llvm/TargetParser/Triple.h:
```diff
@@ -831,6 +831,13 @@ class Triple {
            getArch() == Triple::spirv;
   }
 
+  // Tests whether the target is SPIR-V or SPIR. Currently, we use spir-based
+  // target triples to represent JIT compilation targets for SYCL. We will
+  // transition to using spir-v based target triples to represent JIT
+  // compilation targets for SYCL very soon. This helper function is used
+  // (instead of isSPIR) to ease that transition.
```
Comments that attempt to predict the future have a tendency to either be wrong or to predict the past ;)

tahonermann commented on clang/test/Driver/sycl-offload-jit.cpp:
```diff
@@ -0,0 +1,77 @@
+///
+
+/// Check the phases graph with -fsycl. Use of -fsycl enables offload
+// RUN: %clang -ccc-print-phases --target=x86_64-unknown-linux-gnu \
+// RUN:   -fsycl %s 2>&1 \
+// RUN: | FileCheck -check-prefixes=CHK-PHASES %s
+// RUN: %clang -ccc-print-phases --target=x86_64-unknown-linux-gnu -fsycl \
+// RUN:   %s 2>&1 \
+// RUN: | FileCheck -check-prefixes=CHK-PHASES %s
```
My eyes might be deceiving me, that this looks like the same RUN line written twice with only white space differences.

mdtoguchi replied:
Good catch, will clean up.

tahonermann commented on clang/test/Driver/sycl-offload-jit.cpp:
```diff
@@ -0,0 +1,77 @@
+// RUN: %clang_cl -ccc-print-phases --target=x86_64-pc-windows-msvc -fsycl \
+// RUN:   %s 2>&1 \
+// RUN: | FileCheck -check-prefixes=CHK-PHASES %s
+// CHK-PHASES: 0: input, "[[INPUT:.+\.cpp]]", c++, (host-sycl)
+// CHK-PHASES: 1: preprocessor, {0}, c++-cpp-output, (host-sycl)
+// CHK-PHASES: 2: compiler, {1}, ir, (host-sycl)
```
Hooray for host-side compilation happening first! :)

tahonermann reviewed (commented):
This looks good to me. I have some questions and pointed out some minor concerns.

mdtoguchi committed:
Adjust toolchain files to address a few review comments

  - Improve commenting for unsupported sanitizer options
  - Update visibility of the HostTC, SYCLInstallation and debug info
  - cleanup un-needed header file inclusions

mdtoguchi committed:
Address a few review comments

  - Use spirv32/spirv64 as default triple arch
  - clean up a number of comments
  - Fix file ordering inclusion to be alphabetical

mdtoguchi committed:
Adjust logic for adding options to the -cc1 step

mdtoguchi commented:
Hello - any additional feedback on this?

tschuett commented:
> Hello - any additional feedback on this?

The house rule is to ping:
If you have received no comments on your patch for a week, you can request a review by ‘ping’ing the GitHub PR with “Ping”. The common courtesy ‘ping’ rate is once a week. Please remember that you are asking for valuable time from other professional developers.

https://llvm.org/docs/Contributing.html

mdtoguchi commented:
@bader, thanks for adding the SYCL label, can you also take a look at this PR?

bader commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -767,6 +768,27 @@ Driver::OpenMPRuntimeKind Driver::getOpenMPRuntime(const ArgList &Args) const {
 
+static const char *getDefaultSYCLArch(Compilation &C) {
+  // If -fsycl is supplied we will assume SPIR-V
+  if (C.getDefaultToolChain().getTriple().getArch() == llvm::Triple::x86)
```
If I get it right, here we try to check the address size on the host to propagate the same size for the device compilation.
If so, I suggest we generalize it for more host targets:

```suggestion
  if (C.getDefaultToolChain().getTriple().isArch32Bit())
```

mdtoguchi replied:
Yup, that's right.  I've updated to use your suggestion.

bader commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -4867,12 +4946,13 @@ Action *Driver::ConstructPhaseAction(
-         (Args.hasFlag(options::OPT_fgpu_rdc, options::OPT_fno_gpu_rdc,
-                       false) ||
-          TargetDeviceOffloadKind == Action::OFK_OpenMP))) {
+        (TargetDeviceOffloadKind == Action::OFK_SYCL ||
+         (((Input->getOffloadingToolChain() &&
+            Input->getOffloadingToolChain()->getTriple().isAMDGPU()) ||
+           TargetDeviceOffloadKind == Action::OFK_HIP) &&
+          (Args.hasFlag(options::OPT_fgpu_rdc, options::OPT_fno_gpu_rdc,
+                        false) ||
+           TargetDeviceOffloadKind == Action::OFK_OpenMP)))) {
```
```suggestion
        TargetDeviceOffloadKind == Action::OFK_SYCL ||
        (((Input->getOffloadingToolChain() &&
           Input->getOffloadingToolChain()->getTriple().isAMDGPU()) ||
          TargetDeviceOffloadKind == Action::OFK_HIP) &&
         (Args.hasFlag(options::OPT_fgpu_rdc, options::OPT_fno_gpu_rdc,
                       false) ||
          TargetDeviceOffloadKind == Action::OFK_OpenMP))) {
```

I think we can drop one pair of `()`.

bader reviewed (commented):
@mdtoguchi, thank you for working on this!
Overall, looks good to me, but I made a few suggestions in the comments.
It would be great if @tahonermann and @AaronBallman can confirm that all their comments are addressed.

mdtoguchi committed:
Address a few review items

  - update a number of comments to include SYCL
  - update function names to better fit standards
  - use generic check for 32-bit

mdtoguchi committed:
Update header inclusion and related test

mdtoguchi committed:
Clean up superfluous parens

mdtoguchi commented:
Thanks for the review @bader, I have made updates and hope to have addressed your concerns.

bader reviewed (approved):
Thanks. No other comments from my side.

AaronBallman commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -5116,11 +5123,34 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+      // We want to compile sycl kernels.
+      CmdArgs.push_back("-fsycl-is-device");
+
+      // Set O2 optimization level by default
+      if (!Args.getLastArg(options::OPT_O_Group))
+        CmdArgs.push_back("-O2");
```
I presume this is okay because "debugging" on the device is basically not a thing? Do we expect that debuggers will get support for debugging devices and thus might not want this as a default?

bader replied:
I'm not sure if it's directly related to the "debugging". 
SYCL compiler follows common pattern for GPU programming model implementations, which optimize the offload code by default.
From my POV, more users expect compiler to optimize for speed the code targeting compute accelerator than make code debugger friendly. Is this the right metric for selecting default value from your POV?

BTW, this code has been discussed here too: https://github.com/llvm/llvm-project/pull/107493#discussion_r1763982225.

AaronBallman replied:
Okay, that sounds good to me, thanks!

AaronBallman commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,184 @@
+using namespace llvm::opt;
+
+SYCLInstallationDetector::SYCLInstallationDetector(const Driver &D)
+    : D(D), InstallationCandidates() {
```
```suggestion
    : D(D) {
```
No need

mdtoguchi replied:
Thanks @AaronBallman, updates provided.

mdtoguchi committed:
remove unneeded init code

mdtoguchi committed:
Merge remote-tracking branch 'origin' into sycl-upstream-driver

mdtoguchi committed:
address clang-format findings

mdtoguchi committed:
Merge remote-tracking branch 'origin' into sycl-upstream-driver

mdtoguchi commented:
Ping

AaronBallman reviewed (approved):
LGTM, any concerns from you @MaskRay?

mdtoguchi commented:
Ping

AaronBallman commented:
I reverted the changes due to the failing bots; you can make a new PR with the issues fixed and resubmit for a quick review.

mdtoguchi commented:
Thanks @AaronBallman.  I'll look into it.

tahonermann commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,183 @@
+  SmallString<128> STLWrappersPath(IncludePath);
+  llvm::sys::path::append(STLWrappersPath, "sycl");
+  llvm::sys::path::append(STLWrappersPath, "stl_wrappers");
+  CC1Args.push_back("-internal-isystem");
+  CC1Args.push_back(DriverArgs.MakeArgString(STLWrappersPath));
+  CC1Args.push_back("-internal-isystem");
+  CC1Args.push_back(DriverArgs.MakeArgString(IncludePath));
```
Should these additions be contingent on `--nobuiltininc` having not been specified?

tahonermann reviewed (commented):
I was late to re-review these changes before they landed. However, since the changes have been reverted, I took the opportunity to review again and I have a number of concerns that I would like to see addressed before these changes land again. Though I've noted the issues in this PR, I expect a new PR will be needed to address them and re-land the revised patch.

mdtoguchi commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -0,0 +1,183 @@
+// Unsupported options for SYCL device compilation.
+static std::vector<OptSpecifier> getUnsupportedOpts() {
+  std::vector<OptSpecifier> UnsupportedOpts = {
+      options::OPT_fsanitize_EQ,      // -fsanitize
+      options::OPT_fcf_protection_EQ, // -fcf-protection
+      options::OPT_fprofile_generate,
+      options::OPT_fprofile_generate_EQ,
+      options::OPT_fno_profile_generate, // -f[no-]profile-generate
+      options::OPT_ftest_coverage,
+      options::OPT_fno_test_coverage, // -f[no-]test-coverage
+      options::OPT_fcoverage_mapping,
+      options::OPT_fno_coverage_mapping, // -f[no-]coverage-mapping
+      options::OPT_coverage,             // --coverage
+      options::OPT_fprofile_instr_generate,
+      options::OPT_fprofile_instr_generate_EQ,
+      options::OPT_fno_profile_instr_generate, // -f[no-]profile-instr-generate
+      options::OPT_fprofile_arcs,
+      options::OPT_fno_profile_arcs, // -f[no-]profile-arcs
+      options::OPT_fcreate_profile,  // -fcreate-profile
+      options::OPT_fprofile_instr_use,
+      options::OPT_fprofile_instr_use_EQ,       // -fprofile-instr-use
+      options::OPT_forder_file_instrumentation, // -forder-file-instrumentation
+      options::OPT_fcs_profile_generate,        // -fcs-profile-generate
+      options::OPT_fcs_profile_generate_EQ};
```
@tahonermann, the `InstallationDetector` model was used mimicking the Cuda detectors.  There isn't really an installation to detect here, as everything should be part of the compiler build.  Perhaps the naming is not ideal maybe something more in the lines of `SYCLLocation` or just `SYCLInstallation`?  I'll clean out the `InstallationCandidates` as it is not pertinent.
